### PR TITLE
include/elist: add size() member function support

### PR DESCRIPTION
**Pull request by lxbsz**
It will be very useful if we can get the list size to preallocate
memories for vector to improve the perf.

Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/include/elist.h [hunk 1]
```diff
@@ -32,20 +34,28 @@ template<typename T>
 class elist {
 public:
   struct item {
+  private:
+    elist *_list = NULL;
```
Comment (tchaikov) [inline src/include/elist.h:38]: this increases the size of each element by 8 bytes. not sure it would be a pure win.

File: src/include/elist.h [hunk 2]
```diff
@@ -81,16 +96,21 @@ class elist {
 private:
   item _head;
   size_t item_offset;
+  ceph::atomic<size_t> items = {0};
```
Comment (tchaikov) [inline src/include/elist.h:99]: why `atomic`? using atomic this way enforces the sequential consistency between cores and could cause unnecessary pipeline stall.

Comment (lxbsz) [reply src/include/elist.h:99]: It seems make no sense to use the atomic, the user is responsible to make sure the list is not modified in parallel. I will change it back.

File: src/include/elist.h (hunk 1 above)
Comment (lxbsz) [reply src/include/elist.h:38]: Yeah. Currently only several places will use this and with an alternative we can iterate list and count it in a loop.

Comment (tchaikov) [reply src/include/elist.h:36]: my concern is the scale of the list. what is the size of the linked list in general?

Comment (lxbsz) [reply src/include/elist.h:36]: In mds for some use cases, like the CDir::dirty_dentries, the list size maybe several handreds in overload case, but the list will be trimed every tick.

**Comment by adamemerson** [comment]
Is there any advantage modifying elist gives us that using boost::intrusive::list wouldn't?

**Comment by lxbsz** [comment]
> Is there any advantage modifying elist gives us that using boost::intrusive::list wouldn't?

IMO, one advantage is that the `item->remove_myself()` could simplify the code and with it no need to get to know in which list it is currently in, and it's also actually doing the move. And there has another helper, the `is_on_list() ` could get rid of traversing the list.

**Comment by tchaikov** [comment]
> > Is there any advantage modifying elist gives us that using boost::intrusive::list wouldn't?
> 
> IMO, one advantage is that the `item->remove_myself()` could simplify the code and with it no need to get to know in which list it is currently in, and it's also actually doing the move. And there has another helper, the `is_on_list() ` could get rid of traversing the list.

i think the doubly linked list in boost also has these helpers: https://www.boost.org/doc/libs/1_74_0/doc/html/boost/intrusive/list_base_hook.html

**Comment by lxbsz** [comment]
> > > Is there any advantage modifying elist gives us that using boost::intrusive::list wouldn't?
> > 
> > 
> > IMO, one advantage is that the `item->remove_myself()` could simplify the code and with it no need to get to know in which list it is currently in, and it's also actually doing the move. And there has another helper, the `is_on_list() ` could get rid of traversing the list.
> 
> i think the doubly linked list in boost also has these helpers: https://www.boost.org/doc/libs/1_74_0/doc/html/boost/intrusive/list_base_hook.html

Yeah, it seems. If it work, maybe we could convert some elist in MDS to the intrusive::list ?

**Comment by tchaikov** [comment]
i think this is up to cephfs team. and needs volunteers. as this would involve more efforts, and there are always risks with non-trivial changes.

**Comment by lxbsz** [comment]
> i think this is up to cephfs team. and needs volunteers. as this would involve more efforts, and there are always risks with non-trivial changes.

Yeah, I will try this in late future.

**Review by batrick — COMMENTED** [review]
I think @ukernel has used this predominantly in the MDS for space efficiency reasons. Adding another pointer to `elist::item` is probably too expensive without a clear performance case.

**Comment by batrick** [comment]
@lxbsz suggest we close this?

**Comment by lxbsz** [comment]
> @lxbsz suggest we close this?

Yeah, sure.

**Pull request closed without merging**